proposals/20200909-userpsace-deployment.md

```diff
@@ -0,0 +1,163 @@
+# Falco userspace launch support
+
+## Summary
+
+This is a proposal on how to address deployment of Falco in a userspace
+ environment.
+
+There has been a rise in popularity of serverless architectures that are 
+managed and make kernel instrumentation impossible. This makes it quite hard to
+monitor the runtimes. This project is about defining a patching language/definition
+that allows Falco to be deployed alongside a container in userspace.
+
+## Motivation 
+
+We want to run Falco inside containers without access to kernel space.
+
+By creating a new project (Kilt) - that defines a patching/deployment procedure 
+it is possible to deploy Falco in different userspace environments (k8s, cloud 
+formation, etc)
+
+The project consists of a definition for recipes to include binaries and alter 
+containers, a library to process recipes and runtimes that execute the recipes 
+on supported platforms.
+
+## Goals
+
+- Offer a definition format for patching/deploying Falco along user containers
+- Offer a golang library that parses and executes recipes
+- Define a patching file for Falco
+- Offer automation when using k8s (Admission Controller Mutating Webhook)
+- Offer automation when using cloud formation (CFN Macro)
+
+
+## Non-Goals
+
+- Implementing an image patching runtime
+
+## Proposal
+
+The proposal intends to cover the following use cases: 
+
+- *Existing containers outside of K8S* - This deployment type is the hardest to automate. We 
+will offer instructions on how to alter the containers to add Falco. We can do 
+a migration tool for existing containers but this is not the goal of this
+ project. Adding `patch {}` instructions to the file actually offers a way
+towards offering this functionality as well
+
+- *K8S* - We will offer a K8S Admission Controller Mutating Webhook that will 
+alter pods adding Falco according to patching file
+
+- *K8S-exec* - We will offer a way to execute `patch {}` instructions against 
+specific pods
+
+- *Cloud Formation* - We will offer a cloud formation macro to automatically 
+alter fargate task definitions in a CFN template
+
+
+
+### Components
+- *kilt-cfn* (In PR https://github.com/falcosecurity/evolution/pull/40) - A lambda that patches aws task definitions 
+and can be registered as a macro
+- *kilt-k8s-webhook* (TBD) - A service that listens for webhooks and alters pod
+ resources as they are created
+- *kilt-k8s-exec* - CLI to patch existing k8s pods.
+
+
+## Workflow
+### Preparation (done by Falco maintainers)
+We prepare and publish a docker image containing Falco. The contents would be 
+the following:
+
+```
+/falco/falco - falco binary
+/falco/pdig - ptrace-based instrumentation
+/falco/* - falco configs? other binaries?
+```
+
+We also provide a file called `kilt.cfg` which looks like:
+
+```
```

Comment:
Please annotate the code block with the configuration format you're proposing Radu :)

Also, a sentence on why you do propose this format w.r.t to others could be useful to readers.

Reply:
apparently it's not supported by github :( https://github.com/lightbend/config/issues/543